worker/src/socket.rs

```diff
@@ -0,0 +1,368 @@
+        self: std::pin::Pin<&mut Self>,
+        _cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<std::result::Result<(), std::io::Error>> {
+        // TODO: I don't think we have a flush operation available?
+        Poll::Ready(Ok(()))
+    }
+
+    fn poll_shutdown(
+        mut self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<std::result::Result<(), std::io::Error>> {
+        fn handle_future(
+            cx: &mut std::task::Context<'_>,
+            mut fut: JsFuture,
+        ) -> (Closing, Poll<std::io::Result<()>>) {
+            match fut.poll_unpin(cx) {
+                Poll::Pending => (Closing::Pending(fut), Poll::Pending),
+                Poll::Ready(res) => match res {
+                    Ok(_) => (Closing::None, Poll::Ready(Ok(()))),
+                    Err(e) => (Closing::None, Poll::Ready(Err(js_value_to_std_io_error(e)))),
+                },
+            }
+        }
+        let (new_closing, poll) = match self.close.take().unwrap_or_default() {
+            Closing::None => handle_future(cx, JsFuture::from(self.writer.close())),
+            Closing::Pending(fut) => handle_future(cx, fut),
+        };
+        self.close = Some(new_closing);
+        poll
+    }
+}
+
+pub enum SecureTransport {
+    /// Do not use TLS.
+    Off,
+    /// Use TLS.
+    On,
+    /// Do not use TLS initially, but allow the socket to be upgraded to
+    /// use TLS by calling [`Socket.start_tls`](Socket::start_tls).
+    StartTls,
+}
+
+pub struct SocketOptions {
+    /// Specifies whether or not to use TLS when creating the TCP socket.
+    pub secure_transport: SecureTransport,
+    /// Defines whether the writable side of the TCP socket will automatically
+    /// close on end-of-file (EOF). When set to false, the writable side of the
+    /// TCP socket will automatically close on EOF. When set to true, the
+    /// writable side of the TCP socket will remain open on EOF.
+    pub allow_half_open: bool,
+}
+
+impl Default for SocketOptions {
+    fn default() -> Self {
+        SocketOptions {
+            secure_transport: SecureTransport::Off,
+            allow_half_open: false,
+        }
+    }
+}
+
+/// The host and port that you wish to connect to.
+pub struct SocketAddress {
+    /// The hostname to connect to. Example: `cloudflare.com`.
+    pub hostname: String,
+    /// The port number to connect to. Example: `5432`.
+    pub port: u16,
+}
+
+#[derive(Default)]
+pub struct ConnectionBuilder {
+    hostname: Option<String>,
+    port: Option<u16>,
+    options: SocketOptions,
+}
+
+impl ConnectionBuilder {
+    /// Create a new `ConnectionBuilder` with default settings.
+    pub fn new() -> Self {
+        ConnectionBuilder {
+            hostname: None,
+            port: None,
+            options: SocketOptions::default(),
+        }
+    }
+
+    /// Set the hostname to connect to. Example: `cloudflare.com`.
+    pub fn host(mut self, host: &str) -> Self {
```

Comment:
nit: it might be nice to use `Into<String>` for the host so we accept `String`s as well.

```rust
pub fn host(mut self, host: impl Into<String>) -> Self {
    self.hostname = Some(host.to_string());
    self
}